### PR TITLE
Refactor, add more Rules and use new pycfmodel

### PR DESCRIPTION
**Pull request by jsoucheiron**
### Improvements
- Upgraded to pycfmodel 0.8.1 (this will improve policy action detection)
- Refactored a few classes to use improvements from new base classes and pycfmodel
- `PrivilegeEscalationRule` now detects issues in all policies
### Additions
- New Rules: `SNSTopicDangerousPolicyActionsRule` and `SQSDangerousPolicyActionsRule`
- New abstract base rule: BaseDangerousPolicyActions
### Fixes
- Various typo fixes

### CODE REVIEW

File: cfripper/rules/base_rules.py [hunk 1]
```diff
@@ -122,3 +142,28 @@ def valid_principals(self) -> Set[str]:
...
+    GRANULARITY = RuleGranularity.ACTION
+
+    DANGEROUS_ACTIONS: List[str]
+
+    def resource_invoke(self, resource: Resource, logical_id: str) -> Result:
+        result = Result()
+        for policy in resource.policy_documents:
+            actions = policy.policy_document.get_allowed_actions()
+            dangerous_actions = set(actions) & set(self.DANGEROUS_ACTIONS)
+            self.add_failure_to_result(
```
Comment (oscarbc96) [inline cfripper/rules/base_rules.py:163]: Should we check if dangerous_actions is empty?

**Review by oscarbc96 — COMMENTED** [review]
Very nice!! 🚀

File: cfripper/rules/base_rules.py (hunk 1 above)
Comment (jsoucheiron) [reply cfripper/rules/base_rules.py:163]: Yup, not a bad idea

Comment (jsoucheiron) [reply cfripper/rules/base_rules.py:168]: Done

File: cfripper/rules/base_rules.py [hunk 2]
```diff
@@ -122,3 +142,33 @@ def valid_principals(self) -> Set[str]:
...
                 self._valid_principals.add(self._config.aws_account_id)
         return self._valid_principals
+
+
+class BaseDangerousPolicyActions(ResourceSpecificRule, ABC):
+    """
+    Base class for dangerous actions. Admits a DANGEROUS_ACTIONS class variable with a list of dangerous actions
+    """
+
+    REASON = "Resource {} should not be include the following dangerous actions: {}"
```
Comment (ocrawford555) [inline cfripper/rules/base_rules.py:152]:
```suggestion
    REASON = "Resource {} should not include the following dangerous actions: {}"
```

File: cfripper/rules/sns_topic_policy.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        result = Result()
+        for statement in resource.Properties.PolicyDocument._statement_as_list():
+            if statement.NotPrincipal:
+                self.add_failure_to_result(result, self.REASON.format(logical_id), resource_ids={logical_id})
+        return result
+
+
+class SNSTopicDangerousPolicyActionsRule(BaseDangerousPolicyActions):
+    """
+    Checks for dangerous permissions in Action statements in an SQS Queue Policy.
```
Comment (ocrawford555) [inline cfripper/rules/sns_topic_policy.py:35]:
```suggestion
    Checks for dangerous permissions in Action statements in an SNS Topic Policy.
```

File: cfripper/rules/sns_topic_policy.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    Risk:
+        This is deemed a potential security risk as it could allow privilege escalation.
+    """
+
+    REASON = "SQS Topic policy {} should not not include the following dangerous actions: {}"
```
Comment (ocrawford555) [inline cfripper/rules/sns_topic_policy.py:41]:
```suggestion
    REASON = "SNS Topic policy {} should not not include the following dangerous actions: {}"
```

File: tests/rules/test_PrivilegeEscalationRule.py [hunk 5]
```diff
@@ -1,14 +1,20 @@
...
 
-@pytest.fixture()
+@fixture()
 def valid_role_inline_policy():
     return get_cfmodel_from("rules/PrivilegeEscalationRule/valid_role_inline_policy.json").resolve()
 
 
+@fixture()
+def privilege_escalation_role_cf():
+    print(get_cfmodel_from("rules/PrivilegeEscalationRule/privilege_escalation_role.yaml"))
```
Comment (ocrawford555) [inline tests/rules/test_PrivilegeEscalationRule.py:14]: leftover print

**Review by ocrawford555 — COMMENTED** [review]
Great work Jordi, I only found some small cosmetic changes required! 👍 🚢

**Review by ocrawford555 — APPROVED** [review]

**Review by ocrawford555 — APPROVED** [review]

**Review by oscarbc96 — APPROVED** [review]
🚀

**Pull request merged**